### ISSUE TITLE
Clarity for allowing layout transitions during Transfers of Ownership

### ISSUE THREAD

**Issue by mark-lunarg** [KhronosGroup/Vulkan-Docs#346]
This has been reported as a validation layer issue a couple of times -- in [Resource Sharing Mode](https://www.khronos.org/members/login/sandbox/vulkan/1.0-wsi_extensions/xhtml/vkspec.html#resources-sharing), it is unclear whether a layout transition can be included in the transfer of ownership barriers.  Explicitly stating whether transitions are allowed (or not) would be a big help.

**Comment by krOoze** [comment]
And if so, whether the transition must be identical in both "releasing" and "acquiring" barrier (and if that means single layout transition or two).

**Comment by TobiasHector** [comment]
Ok so you should be allowed to include a layout transfer in the same barrier as an ownership transfer. @krOoze's question is really the tricky one - we need to figure out what was originally meant to happen here, because otherwise combining the two in one barrier is not very useful.

**Comment by krOoze** [comment]
@TobiasHector Well, 1.0.28 seems to answer that (it has to be exactly identical - therefore one transition).

But the rules are somewhat different from regular Barrier making lot of descriptions of it elsewhere a lie. It should probably be described what such barrier pair does exactly (that they act as one barrier with respect to access flags and layout transition and perhaps wheter the transition happens before or after the semaphore or driver can choose).

**Comment by TobiasHector** [comment]
Yes the brief discussion we had indicated that "one transition" was probably the way it was supposed to work, but it's not really written in the spec, and so it's hard to tease out. The spec sort of talks about release and acquire as separate operations, which they are - but they are also apparently meant to split the entire barrier in two, so that one half executes on the first queue, and the other half executes on the other. That's not called out, so some members may not have expected it or implemented it that way. I think we'll discuss this more next week to ensure this is the direction that we intended, and go from there.

**Comment by TobiasHector** [comment]
Oh I forgot to address your last point:

> And perhaps too whether the transition happens before or after the semaphore or driver can choose. By the traditional rules, that's not self-consistent (pretty much still saying it should happen twice based on the src and dst stage of the respective barriers — which is nonsense and would always lead to validation error, because the layout transition parameters must be same in both).

The long-overdue happens-before/happens-after semantics coming to the spec in the next few weeks should sort this mess out - as long as things happen in the right order as defined by the spec (and observed by an application), it doesn't matter what order things go in around that.

I think that what you've said should fall out of the spec quite naturally - they could execute the whole thing on the first queue if they wish, or on the second queue, or split it between two queues - I certainly have no intention of constraining such implementations.

**Comment by krOoze** [comment]
The order is indeed not interesting (at least I can't think of a situation where it would matter).

I was more after the spec self-consistency. There are many places saying something about the Barriers. But that information might not be valid for the case of the queue family transfer, even perhaps after some trivial change to the Resource Sharing chapter (as was done in 1.0.28). E.g. for our layout transition it says:

> When an image memory barrier includes a layout transition, the barrier first makes writes via srcStageMask and srcAccessMask available, then performs the layout transition, then makes the contents of the image subresource(s) in the new layout visible to memory accesses in dstStageMask and dstAccessMask, as if there is an execution and memory dependency between the source masks and the transition, as well as between the transition and the destination masks.

Ok, but we have two sets of masks. So it is supposed to happen twice. Once between the first `srcStage` and the semaphore and once after the semaphore and the second `dstStage`.

That is not just non-obvious or hard to make out from the spec. That is very contradictory to the intended behavior (which is non-obvious and hard to make out in the first place).

**Comment by TobiasHector** [comment]
> The order is indeed not interesting (at least I can't think of a situation where it would matter).

There's potential trickiness with the semaphores, but that should all get figured out.

> I was more after the spec self-consistency.

I plan to move the description of queue ownership transfers to the synchronization section for a start, so we can get all the interactions out in one place. I've already written language to do that, but it interacts with this issue and a couple others so it'll take a little while, but it should get here eventually...

> Ok, but we have two sets of masks. So it is supposed to happen twice. Once between the first srcStage and the semaphore and once after the semaphore and the second dstStage.

Right except it's not really *supposed* to happen twice, because that makes no sense - there is a spec bug here - but need to figure out in what way it is wrong. I've created an internal issue tracking this. It's probably more like you're executing part of the barrier on the first queue, and the other part on the second queue - neither part doing redundant work.

> That is not just non-obvious or hard to make out from the spec. That is very contradictory to the intended behavior (which is non-obvious and hard to make out in the first place).

Like I said - spec bug. I think there's a substantial amount of wording that is outright missing from the spec.

**Comment by TobiasHector** [comment]
So all of this should be clarified in a spec updated due to land in a week or two (which will close this issue) assuming it passes final review, but the gist of what we seem to agree so far is this:

Whilst two barriers are executed, the layout transition (which must be specified *identically* in *both barriers*) only executes *once*. It executes at some point after the release barrier's availability operations and before the acquire barrier's visibility operations. It could happen entirely on either queue, or partially on both; it just has to happen before you ask for it to be useful (the acquire barrier).

I don't want to *completely* commit that this is the right answer because we still need the WG to do a final sign-off to put it in the spec, but the issue was raised and discussed in great detail to this get this conclusion already, and I don't expect it to change now.

**Comment by oddhack** [comment]
This should be fixed in the 1.0.35 spec update, as part of the synchronization language rewrite.

**Comment by ParticlePeter** [comment]
Well defined now. Just few comments (reading the pdf version v1.0.35):
1. I recommend adding a hyper link from chapter 11.7 (last two paragraphs about ownership transfer) to chapter 6.6.4
2. The enumerations in chapter 6.6.4 are misaligned (most probably this was pointed out already) right after: 
> A queue family ownership transfer consists of two distinct parts:

**Comment by TobiasHector** [comment]
> I recommend adding a hyper link from chapter 11.7 (last two paragraphs about ownership transfer) to chapter 6.6.4

There's already such a link if you click "queue ownership transfer". I guess we could say something like "(see 6.6.4 for more information)" to be more explicit, but given that the details of that operation are covered by a link in the one place it's mentioned, I'm not really sure that is necessary. Where precisely are you thinking we should add another link?

> The enumerations in chapter 6.6.4 are misaligned (most probably this was pointed out already) right after:

Ack yea I noticed this and cleaned it up but it crept back in... ohhh I wonder, we have a reflow script which makes sure we always start a new sentence on a new line, and I bet it's bugging out here thinking "1." is a sentence. Blah, I'll fix that.

**Comment by TobiasHector** [comment]
NB: After a bug has been closed, we don't appear to get notifications about messages that have been posted. If you've got other feedback then either re-open the bug, or start a new one. In this case I just happened to see your comment so I'll fix it, but I won't have an external bug to notify you when it is fixed at present. In fact you might not even see this - I can but hope :S

**Comment by krOoze** [comment]
@TobiasHector You can @ someone if you want them to be notified.
Hmm this makes reopening Issue a problem. I kinda hoped Khronos members have the repo watched (you then get notification of all the comunication). Or at lest one member that can then delegate it if need be.

**Comment by TobiasHector** [comment]
It looks like I was having problems with my email this morning which I hadn't noticed - sorted now. I did actually get an email for that comment after all.

Even in lieu of that, if the bug is reopened, it shows up in the list of open bugs, which we actively go through (so if we miss something, if it's in an open bug it'll get fixed at *some* point). If it's in the closed list and we've missed it, there's no safety net for that. So would still rather reopen bugs if there's work to do.

Regardless, the list thing will be fixed in the next version, just raised an internal MR which should be non-contentious, which just leaves the question about the link.

**Comment by krOoze** [comment]
> If it's in the closed list and we've missed it, there's no safety net for that. So would still rather reopen bugs if there's work to do.

Good to know. Already re-newed one Issue in that spirit.

**Comment by TobiasHector** [comment]
> (FYI we cannot reopen if someone else closes an Issue).

Oh weird. I can, but I guess that's because I'm part of the KhronosGroup organization and you're not...? I suppose that complicates things - though probably best to raise a new issue anyway; in theory that stops the conversation getting confused.

**Comment by krOoze** [comment]
Yes, because that. GitHub has weird design choices sometimes.

**Comment by ParticlePeter** [comment]
@TobiasHector 

> There's already such a link if you click "queue ownership transfer". I guess we could say something like "(see 6.6.4 for more information)" to be more explicit, but given that the details of that operation are covered by a link in the one place it's mentioned, I'm not really sure that is necessary. Where precisely are you thinking we should add another link?

I tried to hover over all the text but somehow missed that one. Seems a good conceptual place for it but visually its not very noticeable compared to the web version. This is the case for lot of the links, have you considered to maybe color them differently in the pdf version as well?

**Comment by TobiasHector** [comment]
Huh. Didn't realise that the pdf version wasn't highlighting links in any way, that seems... annoying. When we finally move to Asciidoctor this should be taken care of, but in the meantime I guess we could ask @oddhack if he knows a way to highlight them...?
